data/blog/enjoyable-database-testing.md

```diff
@@ -0,0 +1,277 @@
+
+To limit the overhead starting a ephemeral database it is best to create the minimal number of database clusters.
+
+`hspec` is missing an `aroundAll` function. Here is one that I use: [aroundAll](https://gist.github.com/jfischoff/5cf62b82e1dd7d03c8a610ef7fd933ff)
+
+We can use it to replace the `around` in our previous example:
+
+```haskell
+aroundAll withSetup $ describe "list/add/delete" $ do
+  it "return [] initially" $ withPool $ \conn -> do
+    list conn `shouldReturn` []
+  it "returns the created elements" $ withPool $ \conn -> do
+    theId <- add conn 1
+    list conn `shouldReturn` [theId]
+  it "deletes what is created" $ withPool $ \conn -> do
+    theId <- add conn 2
+    delete conn theId
+    list conn `shouldReturn` []
+```
+
+## The Rub
+
+Using `aroundAll` will speed up our testing but our test suite is now broken. The problem is the second test leaves behind
+an entry breaking the third test.
+
+We could make the third test more robust by caching the state at the start of the test and ensuring we return to the initial state:
+
+```haskell
+  it "deletes what is created" $ withPool $ \conn -> do
+    before <- list conn
+
+    theId <- add conn 2
+    delete conn theId
+    list conn `shouldReturn` before
+```
+
+There are still potential problems with this modification.
+
+At the end of the day it might not be clear how to make the test robust.
+
+One way we can regain the isolation of separate database clusters and `postgres` instances is by using the databases mechanisms for query isolation.
+
+To faciliate query isolation we'll write an function to wrap a list of statements in a transaction that we rollback instead of committing:
+
+```haskell
+abort :: (Connection -> IO a) -> Connection -> IO a
+abort f conn = bracket_
+  (execute_ conn "BEGIN")
+  (execute_ conn "ROLLBACK")
+  (f conn)
+```
+
+We can now prefix our tests with abort and they will not interfer with each other
+
+```haskell
+aroundAll withSetup $ describe "list/add/delete" $ do
+  it "return [] initially" $ withPool $ abort $ \conn -> do
+    list conn `shouldReturn` []
+  it "returns the created elements" $ withPool $ abort $ \conn -> do
+    theId <- add conn 1
+    list conn `shouldReturn` [theId]
+  it "deletes what is created" $ withPool $ abort $ \conn -> do
+    theId <- add conn 2
+    delete conn theId
+    list conn `shouldReturn` []
+```
+
+It is worth pointing out that not every PostgreSQL sql statement can run in a transaction. Additionally there is no isolation level that can bring the database to the same state the way starting at a cluster snapshot can. Things like series are incremented and not decrememented on rollback among other MVCC infidelities. That said the fast majority of queries are perfectly isolated using  transaction and `abort` will be sufficent to return the database to a original state (for all intensive purposes).
```

Comment:
> using  transaction

extra space

Reply:
> That said the fast majority of queries are perfectly isolated using  transaction and `abort` will be sufficent to return the database to a original state (for all intensive purposes).

That said the fast majority `abort` is perfectly sensible solution for testing most queries.